cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -247,4 +269,34 @@ public WorkflowToken deepCopy() {
     }
     return builder.build();
   }
+
+  private void ensureEnoughMemoryAvailable(String key, Value value) {
+    int left = bytesLeft;
+    left -= key.length();
+    if (left <= 0) {
+      throw new IllegalStateException(String.format("Exceeded maximum permitted size of workflow token '%sMB' while " +
+                                                      "adding key '%s'. Current size is '%sMB'. Please increase the " +
+                                                      "maximum permitted size by setting the parameter '%s' in " +
+                                                      "cdap-site.xml to add more values.",
+                                                    maxSizeBytes / (1024 * 1024), key,
+                                                    currentSizeBytes / (1024 * 1024),
+                                                    Constants.AppFabric.WORKFLOW_TOKEN_MAX_SIZE_MB));
+    }
+    left -= value.toString().length();
+    if (left <= 0) {
+      throw new IllegalStateException(String.format("Exceeded maximum permitted size of workflow token '%sMB' while " +
```

Comment:
You can share this method between merge and put easily, without repeating the contract of append (which is check size, add, update size). You may modify the signature a bit to take a `List<NodeValue>` instead of a `Map`.